google/cloud/bigtable/batcher.py

```diff
@@ -125,22 +298,120 @@ def mutate_rows(self, rows):
+            rows_to_flush.append(row)
+            batch_info.mutations_count = mutations_count
+            batch_info.rows_count = rows_count
+            batch_info.mutations_size = mutations_size
+
+            if (
+                rows_count >= self.flush_count
+                or mutations_size >= self.max_row_bytes
+                or mutations_count >= self.flow_control.max_mutations
+                or mutations_size >= self.flow_control.max_row_bytes
+                or self._rows.empty()  # submit when it reached the end of the queue
+            ):
+                self.flow_control.control_flow(batch_info)
+                future = self._executor.submit(self.flush_rows, rows_to_flush)
+                self.futures_mapping[future] = batch_info
+                future.add_done_callback(self.batch_completed_callback)
+
+                # reset and start a new batch
+                rows_to_flush = []
+                mutations_size = 0
+                rows_count = 0
+                mutations_count = 0
+                batch_info = BatchInfo()
+
+    def batch_completed_callback(self, future):
+        """Callback for when the mutation has finished.
+
+        Raise exceptions if there's any.
+        Release the resources locked by the flow control and allow enqueued tasks to be run.
+        """
+
+        processed_rows = self.futures_mapping[future]
+        self.flow_control.release(processed_rows)
+        del self.futures_mapping[future]
+
+    def flush_rows(self, rows_to_flush=None):
+        """Mutate the specified rows.
+
+        raises:
+            * :exc:`.batcherMutationsBatchError` if there's any error in the
+                mutations.
+        """
+        responses = []
+        if len(rows_to_flush) > 0:
+            response = self.table.mutate_rows(rows_to_flush)
+
+            for result in response:
+                if result.code != 0:
+                    exc = from_grpc_status(result.code, result.message)
+                    self.exceptions.put(exc)
+                responses.append(result)
+
+        return responses
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        """Clean up resources. Flush and shutdown the ThreadPoolExecutor."""
+        self.close()
+
+    def close(self):
+        """Clean up resources. Flush and shutdown the ThreadPoolExecutor.
+        Any errors will be raised.
+
+        raises:
+            * :exc:`.batcherMutationsBatchError` if there's any error in the
+                mutations.
+
         """
-        if len(self.rows) != 0:
-            self.table.mutate_rows(self.rows)
-            self.total_mutation_count = 0
-            self.total_size = 0
-            self.rows = []
+        self._is_open = False
+        self.flush()
+        self._executor.shutdown(wait=True)
+        if self.exceptions.qsize() > 0:
+            exc = list(self.exceptions.queue)
+            raise MutationsBatchError("Errors in batch mutations.", exc=exc)
```

Comment:
Maybe also unregister the `atexit` callback registered in `__init__`?

Reply:
Yes, added.